multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,303 @@
+	}
+}
+
+// +kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=resourceexports,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+func (r *LabelIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Reconciling Pod/Namespace for label identity", "pod/ns", req.NamespacedName)
+	var commonArea commonarea.RemoteCommonArea
+	var err error
+	commonArea, r.localClusterID, err = r.commonAreaGetter.GetRemoteCommonAreaAndLocalID()
+	if commonArea == nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	r.remoteCommonArea = commonArea
+	var pod v1.Pod
+	var ns v1.Namespace
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).InfoS("Pod is deleted", "pod", req.NamespacedName)
+			defer r.labelMutex.Unlock()
+			return r.removeLabelForPod(ctx, req.NamespacedName.String())
+		}
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: req.Namespace}, &ns); err != nil {
+		klog.ErrorS(err, "Cannot get corresponding Namespace of the Pod", "pod", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	normalizedLabel := getNormalizedLabel(ns.Labels, pod.Labels, ns.Name)
+	return r.onPodCreateOrUpdate(ctx, req.NamespacedName.String(), normalizedLabel)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *LabelIdentityReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Pod{}).
+		WithEventFilter(predicate.LabelChangedPredicate{}).
+		Watches(&source.Kind{Type: &v1.Namespace{}}, handler.EnqueueRequestsFromMapFunc(r.namespaceMapFunc)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
+		Complete(r)
+}
+
+// namespaceMapFunc handles Namespace update events (Namespace label change) by enqueuing
+// all Pods in the Namespace into the reconciler processing queue.
+func (r *LabelIdentityReconciler) namespaceMapFunc(ns client.Object) []reconcile.Request {
+	podList := &v1.PodList{}
+	r.Client.List(context.TODO(), podList, client.InNamespace(ns.GetName()))
+	requests := make([]reconcile.Request, len(podList.Items))
+	for i, pod := range podList.Items {
+		requests[i] = reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      pod.GetName(),
+				Namespace: pod.GetNamespace(),
+			},
+		}
+	}
+	return requests
+}
+
+// This function needs to be called with labelMutex acquired.
+func (r *LabelIdentityReconciler) setLabelExportUpdatesInProgress(label string) {
+	r.labelExportUpdatesInProgress.Insert(label)
+	// Unlock the labelMutex to allow cache updates for other labels. Any worker operating on
+	// the label will be blocked as it is added into labelExportUpdateInProgress.
+	r.labelMutex.Unlock()
```

Comment:
I feel it is hard to track where the mutex is locked/unlocked, if lock/unlock is not in the same func. Do we really need to do this?

Reply:
+1
Could we do the following?
1. `Reconcile` calculates a Pod's normalized label and maintains mappings between pods and normalized labels, but doesn't really create/delete labelExport and just enqueues the old and new labels to a queue of labels.
2. A separate pool of workers create or delete labelExport based on whether the label is used by any Pod.

There would be no race conditions when a label is used by multiple Pods, and concurrent label creation/deletion is implemented lock-free.
The only critical section is when `Reconcile` updates the mappings with write Lock and the label workers read the mappings with read lock.

Reply:
Done